streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java

```diff
@@ -0,0 +1,365 @@
+
+    InMemorySessionStore(final String name,
+                         final long retentionPeriod,
+                         final String metricScope) {
+        this.name = name;
+        this.retentionPeriod = retentionPeriod;
+        this.metricScope = metricScope;
+
+        this.openIterators = ConcurrentHashMap.newKeySet();
+        this.endTimeMap = new ConcurrentSkipListMap<>();
+    }
+
+    @Override
+    public String name() {
+        return this.name;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void init(final ProcessorContext context, final StateStore root) {
+        this.context = (InternalProcessorContext) context;
+
+        final StreamsMetricsImpl metrics = this.context.metrics();
+        final String taskName = context.taskId().toString();
+        expiredRecordSensor = metrics.storeLevelSensor(
+            taskName,
+            name(),
+            EXPIRED_WINDOW_RECORD_DROP,
+            Sensor.RecordingLevel.INFO
+        );
+        addInvocationRateAndCount(
+            expiredRecordSensor,
+            "stream-" + metricScope + "-metrics",
+            metrics.tagMap("task-id", taskName, metricScope + "-id", name()),
+            EXPIRED_WINDOW_RECORD_DROP
+        );
+
+        if (root != null) {
+            context.register(root, (key, value) -> {
+                put(SessionKeySchema.from(Bytes.wrap(key)), value);
+            });
+        }
+        this.open = true;
+    }
+
+    @Override
+    public void put(final Windowed<Bytes> sessionKey, final byte[] aggregate) {
+        removeExpiredSegments();
+
+        final long windowEndTimestamp = sessionKey.window().end();
+        observedStreamTime = Math.max(observedStreamTime, windowEndTimestamp);
+
+        if (windowEndTimestamp <= observedStreamTime - retentionPeriod) {
+            expiredRecordSensor.record();
+            LOG.debug("Skipping record for expired segment.");
+        } else {
+            if (aggregate != null) {
+                endTimeMap.computeIfAbsent(windowEndTimestamp, t -> new ConcurrentSkipListMap<>());
+                final ConcurrentNavigableMap<Bytes, ConcurrentNavigableMap<Long, byte[]>> keyMap = endTimeMap.get(windowEndTimestamp);
+                keyMap.computeIfAbsent(sessionKey.key(), t -> new ConcurrentSkipListMap<>());
+                keyMap.get(sessionKey.key()).put(sessionKey.window().start(), aggregate);
+            } else {
+                remove(sessionKey);
+            }
+        }
+    }
+
+    @Override
+    public void remove(final Windowed<Bytes> sessionKey) {
+        final ConcurrentNavigableMap<Bytes, ConcurrentNavigableMap<Long, byte[]>> keyMap = endTimeMap.get(sessionKey.window().end());
+        final ConcurrentNavigableMap<Long, byte[]> startTimeMap = keyMap.get(sessionKey.key());
+        startTimeMap.remove(sessionKey.window().start());
+
+        if (startTimeMap.isEmpty()) {
+            keyMap.remove(sessionKey.key());
+            if (endTimeMap.isEmpty()) {
```

Comment:
Not sure what this is doing. Seems like it should be:
```
if (keyMap.isEmpty()) {
    endTimeMap.remove(sessionKey.window().end());
}
```

Reply:
Hm, not sure what happened here either. Thanks for catching this, will fix